discussion/rust-for-embedded-systems.tex

```diff
@@ -0,0 +1,90 @@
+It provides some of the same functionality that we implemented for the event-hub described in \autoref{sec:irq-closures}.
+We can not use this project directly because it depends on {\std}, but if we look closer at the project's source code, we can see that the project author has implemented an alternate (and outdated) version which is only dependent on {\core}.
+Another example is the \texttt{smallvec}\footnote{\url{https://crates.io/crates/smallvec}} project, which provides a handful of optimized versions of {\rust}'s vector structure that have a length-limit of 32 entries.
+Similar to the \texttt{lazy-static} project, all the dependencies for \texttt{smallvec} are included from {\std}, but are also available in {\core}.
+We have implemented a structure that is similar to \texttt{smallvec} as part of the {\tracker} application, but it feels like an unnecessary addition because it is already available on \texttt{crates.io} \emph{and} it only depends on features from {\core}.
+
+This distinction between {\std} and \gls{rsl} feels somewhat contradictory to {\rust}'s focus on modularization and package distribution with {\cargo}.
+It also feels destructive for non-standard projects like ours, because we would either have to re-implement or modify the already existing projects in order to get them to work for our platform.
+We hope that it will get easier in time to define non-standard projects, while simultaneously being able to utilize many of the great open-source libraries that are already available.
+
+\subsection{Language Challenges}
+\label{sec:disc:lang-challenges}
+
+In \autoref{sec:intro:assignment} we identified six language challenges that must be considered when using a language in an embedded system.
+Here, we revisit each challenge and discuss how they were solved for the {\rg} platform.
+
+\begin{description}
+\item [LC1 {\lci}] \hfill \\
+  {\rust} exposes two \concept{intrinsics} for handling volatile read and write.
+  These makes the code more verbose compared to the mechanism in {\C}, where a variable is marked with the \code{volatile} keyword.
+  In {\rust}, the intrinsic functions must be used each time a variable is read or written.
+  This, however, gives the programmer more fine-grained control as a variable can be used both as volatile and non-volatile.
+
+\item [LC2 {\lcii}] \hfill \\
+  Interrupt handlers must use the {\C} \gls{abi}, because the interrupts are dispatched from the {\C} runtime.
+  This makes the code more verbose, but it is easily achievable in {\rust}.
+  As discussed in \autoref{sec:avoiding_mutable_aliases_to_hardware}, a downside to the interrupted programming model found in embedded programming for {\rust} is the reliance on global mutable state.
+  In these circumstances, the compiler is limited in its ability to statically verify safety because accesses to these variables must be contained inside {\unsafe} blocks.
```

Comment:
Append: `The Zinc.rs project makes interrupt handlers native to {/rust} by replacing the {\C} runtime, this however does not tackle the problem of global mutable state.`